### ISSUE TITLE
Recall over precision for blocking phase.

### ISSUE THREAD

**Issue by jdvala** [dedupeio/dedupe#898]
Hi folks, 

I have run the algorithm on my dataset and observed that I can attain high precision in the blocking phase (~ `0.97`) and the recall for the same is around `0.6` . I was wondering if there is any way to compromise on precision for recall to be higher.

What I do currently,

* Create the data dictionary as done [here](https://github.com/dedupeio/dedupe-examples/blob/e09cac9e09b1534508a064197f6f86275fe9fa87/csv_example/csv_example.py#L117 ) like `data_dict = create_data_dict(dataframe)`
* Initialize the dedup object with variables like [here](https://github.com/dedupeio/dedupe-examples/blob/e09cac9e09b1534508a064197f6f86275fe9fa87/csv_example/csv_example.py#L99) `deduplication = dedupe.Dedupe(fields)`
* Prepare training `deduplication.prepare_training(data_dict)`
* Console labeling `dedupe.console_label(deduplication)`
* Train the algorithm.`deduplication.train()`
* Get the pairs using `pairs = deduplication.pairs(data_dict)` with the data dict created in the first step
* Score the pairs using `scores = deduplication.score(pairs)` 
* Select a score threshold and compare it with my true labels.

It would be really good if I can tweak the `deduplication.score` function to consider more pairs to increase the recall.

**Comment by fgregg** [comment]
i've been thinking about this problem for a while, and i it's tricky.

one dial that we could turn is which records we use for the final block training. 

right now, we only use the records pairs that have been labeled as a match. we could do something else and use all the records from the training set that the classifier believes to has above some X% threshold of being a match.

unfortunately, there are no real dials to turn for you right now except to add more training data.

**Comment by jdvala** [comment]
Hi @fgregg,

Thanks for the reply, I will take a look at how the algorithm works in details now to see how might I be able to achieve it and from your answer I don't know if I should.

Here are my thoughts, which I want you to take with a pinch of salt. 

The `dedupliaction.pairs` if they return all the pairs no matter if they met a certain criteria or not, could potentially make it easy for score function to score them. I believe that internally it does this however based on some condition it might decide to select only a portion of the pairs and then send it to score. If this is the case may be there is an internal method which can give all the pairs, that can then be converted to generator like `pairs` object and then given to score for scoring. I don't know if this is possible

**Comment by fgregg** [comment]
you could certainly override `pairs` to do that. this will certainly increase recall, but at the cost of comparing all N * N-1/2 records, which will not be practicable if you have even a medium-sized dataset

**Comment by jdvala** [comment]
Certainly it will not be practical to have `N * (N-1)/2` pairs created. However, I can filter these pairs based on simple condition before passing it to scoring method. There is a column in my dataset that can help filter out around 40-50% of pairs. Thanks for all the help.

**Comment by jdvala** [comment]
Closing the issue as I have found my solution. Thank you @fgregg for the support
